### PR TITLE
more fixes for amr.subcycling_mode = None + retry

### PR DESCRIPTION
**Pull request by zingale**
## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

File: Source/driver/Castro_advance_ctu.cpp [hunk 1]
```diff
@@ -157,11 +157,18 @@ Castro::retry_advance_ctu(Real dt, const advance_status& status)
...
+            max_level_to_advance = parent->finestLevel();
         }
-        else {
-            dt_subcycle = std::min(dt, dt_subcycle) * retry_subcycle_factor;
+
+        for (int lev = level; lev <= max_level_to_advance; ++lev) {
+            if (status.suggested_dt > 0.0_rt && status.suggested_dt < dt) {
+                getLevel(lev).dt_subcycle = status.suggested_dt;
+            } else {
+                getLevel(lev).dt_subcycle = std::min(dt, dt_subcycle) * retry_subcycle_factor;
```
Comment (yut23) [inline Source/driver/Castro_advance_ctu.cpp:170]: This was setting `dt_subcycle` for the finer levels to be different from the value for level 0. It doesn't appear to fix the negative densities, though.
```suggestion
                getLevel(lev).dt_subcycle = std::min(dt, getLevel(lev).dt_subcycle) * retry_subcycle_factor;
```

**Comment by yut23** [comment]
Is this bite taken out of the secondary an issue? inputs_2d_collision.test says it sets `problem.collision_separation = 1.25` to start with the stars already merging, but these don't look like very well-behaved initial conditions.
Screenshot from 2024-04-18 19-11-08

**Comment by zingale** [comment]
if you build Sedov in 2D and run as:

```
./Castro2d.gnu.MPI.ex inputs.2d.sph_in_cylcoords castro.change_max=1.01 amr.subcycling_mode=None castro.fixed_dt=2.e-5
```
then it retries after step 2 and dies in the Riemann solver

**Comment by zingale** [comment]
actually, it runs if we add `castro.riemann_solver=0`

**Comment by zingale** [comment]
okay... with this PR, the Sedov retries a bit at the start and then runs fine.  Looking at the output, it looks reasonable (by eye).  I will compare to the subcycling case in a bit.

On development, it constantly retries and then dies with too many subcycles.  

So I think that these fixes are needed.

**Comment by yut23** [comment]
This number (which I'm assuming is some sort of placeholder) seems to be leaking into the state variables. Changing it to 1.e40 changes all the invalid density values, for example from -5e+65 to -5e+45. https://github.com/AMReX-Astro/Castro/blob/5f44c133d85369cba7d2a12edc5f40ffaa271b3e/Source/gravity/Gravity.cpp#L396

File: Source/driver/Castro_advance_ctu.cpp [hunk 2]
```diff
@@ -331,19 +346,18 @@ Castro::subcycle_advance_ctu(const Real time, const Real dt, int amr_iteration,
...
-            swap_state_time_levels(0.0);
+            for (int lev = level; lev <= max_level_to_advance; ++lev) {
+                getLevel(lev).swap_state_time_levels(0.0);
 
 #ifdef GRAVITY
-            if (do_grav) {
-                gravity->swapTimeLevels(level);
-            }
+                if (do_grav) {
+                    getLevel(lev).gravity->swapTimeLevels(level);
```
Comment (yut23) [inline Source/driver/Castro_advance_ctu.cpp:354]: Found it.
```suggestion
                    getLevel(lev).gravity->swapTimeLevels(lev);
```

**Review by yut23 — APPROVED** [review]

**Pull request merged**